lib/src/lexer.c

```diff
@@ -239,23 +289,36 @@ static void ts_lexer__mark_end(TSLexer *_self) {
 static uint32_t ts_lexer__get_column(TSLexer *_self) {
   Lexer *self = (Lexer *)_self;
   
-  uint32_t goal_byte = self->current_position.bytes;
-  
   self->did_get_column = true;
-  self->current_position.bytes -= self->current_position.extent.column;
-  self->current_position.extent.column = 0;
-
-  if (self->current_position.bytes < self->chunk_start) {
-    ts_lexer__get_chunk(self);
-  }
 
   uint32_t result = 0;
-  ts_lexer__get_lookahead(self);
-  while (self->current_position.bytes < goal_byte && !ts_lexer__eof(_self) && self->chunk) {
-    ts_lexer__do_advance(self, false);
-    result++;
-  }
+  if (!ts_lexer__try_get_cached_column_value(self, &result)) {
+    uint32_t goal_byte = self->current_position.bytes;
+    
+    self->current_position.bytes -= self->current_position.extent.column;
+    self->current_position.extent.column = 0;
 
+    if (self->current_position.bytes < self->chunk_start) {
+      ts_lexer__get_chunk(self);
+    }
+
+    result = 0;
+    ts_lexer__get_lookahead(self);
+    
+    // Don't count BOM in column position
+    if (
+      self->current_position.bytes == 0 &&
+      self->data.lookahead == BYTE_ORDER_MARK
+    ) ts_lexer__do_advance(self, false, false);
+
+    while (self->current_position.bytes < goal_byte && !ts_lexer__eof(_self) && self->chunk) {
+      ts_lexer__do_advance(self, false, false);
+      result++;
+    }
+
+    ts_lexer__set_column_cache(self, result);
+  }
```

Comment:
I think that you simplify this code a little bit. I would suggest:

1. Eliminating the `update_column_cache` parameter to `ts_lexer__do_advance`. Just have that method *always* update the column cache:

2. Restructure this `ts_lexer__get_column` function so that it just populates the column cache if necessary, using the normal code path, and then returns its value:

```c
static uint32_t ts_lexer__get_column(TSLexer *_self) {
  Lexer *self = (Lexer *)_self;
  self->did_get_column = true;

  if (!self->column_cache.valid) {
    uint32_t goal_byte = self->current_position.bytes;
    self->current_position.bytes -= self->current_position.extent.column;
    self->current_position.extent.column = 0;
    self->column_cache.valid = true;
    self->column_cache.value = 0;

    // ... (advance back to the initial position)
  }
  
  return self->column_cache.value;
}
```